cocos2d/CCImage.h

```diff
@@ -0,0 +1,149 @@
+
+ @since 4.0
+ */
+extern NSString * const CCImageOptionRescaleFactor;
+
+/**
+ Should the image dimensions be expanded to a power of two while loading?
+ 
+ The default value is `NO`.
+
+ @since 4.0
+ */
+extern NSString * const CCImageOptionExpandToPOT;
+
+/**
+ Should the image alpha be premultiplied while loading?
+ 
+ The default value is YES.
+ 
+ @warning Some image loaders only support pre-multiplied alpha.
+
+ @since 4.0
+ */
+extern NSString * const CCImageOptionPremultiply;
+
+/**
+ Low level image handling class for RGBA8 bitmaps.
+ */
+@interface CCImage : NSObject<NSCoding>
+
+/**
+ Initialize a new image with raw pixel data. All default options are applied to the image.
+ 
+ TODO Should this initializer support an options dictionary?
+ TODO Should `pixelData` be allowed to be `nil`?
+
+ @param pixelSize    Size of the image in pixels.
+ @param contentScale Content scale of the image.
+ @param pixelData    A pointer to raw, tightly packed, RGBA8 pixel data.
+
+ @return An image object that wraps the given pixel data.
+
+ @since 4.0
+ */
+-(instancetype)initWithPixelSize:(CGSize)pixelSize contentScale:(CGFloat)contentScale pixelData:(NSMutableData *)pixelData;
+
+/**
+ Initialize a new image from a CGImageRef.
+
+ @param image        The CGImage to use as the image's content.
+ @param contentScale The content scale the CGImage should be interpreted as.
+ @param options      A dictionary of NSImageOption* key and NSNumbers for values.
+ May be `nil`. Any keys not included will be filled with default values.
+
+ @return An image with the CGImage loaded into it.
+
+ @since 4.0
+ */
+-(instancetype)initWithCGImage:(CGImageRef)image contentScale:(CGFloat)contentScale options:(NSDictionary *)options;
+
+/**
+ Initialize an image based on a CCFileUtils file object.
+
+ @param file    The CCFile to load the image data from.
+ @param options A dictionary of NSImageOption* key and NSNumbers for values.
+ May be `nil`. Any keys not included will be filled with default values.
+
+ @return An image loaded from the file.
+
+ @since 4.0
+ */
+-(instancetype)initWithCCFile:(CCFile *)file options:(NSDictionary *)options;
+
+/**
+ Size of the image's bitmap in pixels.
+
+ @since 4.0
+ */
+@property(nonatomic, readonly) CGSize sizeInPixels;
+
+/**
+ Bitmap data pointer. The format will always be RGBA8.
+
+ @since 4.0
+ */
+@property(nonatomic, readonly) NSMutableData *pixelData;
+
+/**
+ Content scale of the bitmap
+
+ @since 4.0
+ */
+@property(nonatomic, readonly) CGFloat contentScale;
+
+/**
+ User assignable content size of the image in points.
+ 
+ This value may not equal pixelSize/contentScale if the image is padded.
+ It defaults to the original size of the image in points.
+
+ @since 4.0
+ */
+@property(nonatomic, assign) CGSize contentSize;
+
+
+// TODO Other thoughts:
+// * Method to create a CGImage from the CCImage?
+// * Method to capture the screen as a CCImage?
+
```

Comment:
Hmm. Since we are refactoring the director/view stuff. Maybe it would make sense to put an instance method on the view class that grabs the view's pixels as a CCImage. That might make more sense.